File: Python/Trees.py
class Node:
    data=0
    left=None
    right=None
    def __init__(self,val,left,right):
        self.data = val
        self.left = left
        self.right=right
    def __init__(self,val):
        self.data = val


def levelOrder(root):
    que = list()
    que.append(root)
    while(que):
        for i in range(0,len(que)):
            curr = que.pop(0)
            print(curr.data,end=" ")
            if curr.left is not None: que.append(curr.left)
            if curr.right is not None: que.append(curr.right)
        print()
                    


def insert(root,val):
    if(root==None):
        return Node(val)
    
    if(val < root.data):
        root.left = insert(root.left,val)
    elif val>=root.data:
        root.right = insert(root.right,val)
    
    return root

def display(root):
    if(root==None): return
    
    display(root.left)
    print(root.data,end=" ")
    display(root.right)

root = None
while True:
    x = int(input())
    if(x==-1): break
    root = insert(root,x)

display(root)
print()
levelOrder(root)

File: Practice/C_Practice/Trees.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>


typedef struct Node{
    int data;
    Node* left;
    Node* right;
} Node;


Node* q[100];
int top=0,tail=-1;

void push(Node* n){
    q[++tail] = n;
}

int empty(){
    return (top>tail);
}

Node* pop(){
    if(top>tail){
        return  NULL;
    }
    return q[top++];
}

Node* peek(){
    return q[top];
}

Node* insert(Node* root,int val){
    if(root==NULL){
        Node* newnode = (Node*)malloc(sizeof(Node));
        newnode->data = val;
        return newnode;
    }
    if(val<root->data){
        root->left = insert(root->left,val);
    }
    else{
        root->right = insert(root->right,val);
    }
    return root;
}

void display(Node* root){
    if(root==NULL) return;
    display(root->left);
    printf("%d ",root->data);
    display(root->right);
}

Node* smallest(Node* root){
    if(root==NULL) return root;
    return(root->left);
}

void lv_order(Node* root){
    if(root==NULL) return;
    push(root);
    while(!empty()){
        int lvl = tail-top +1;
        for(int i=0;i<lvl;i++){
            Node* curr = pop();
            if(curr!=NULL){
                printf("%d ",curr->data);
                if(curr->left != NULL) push(curr->left);
                if(curr->right != NULL) push(curr->right);
            }
        }
        printf("\n");
    }
}


int main(){
    Node* root = NULL;
    int x = 0;
    while(empty()){
        scanf("%d",&x);
        if(x==-1) break;
        root = insert(root,x);
    }
    // display(root);
    printf("Level order : \n");
    lv_order(root);
    Node* small = smallest(root);
    printf("Smallest Node in the tree is : %d located at %a",small->data,small);
    return 0;
}